core/src/wallet/ethereum/ERC20/ERC20LikeAccount.cpp

```diff
@@ -74,15 +75,95 @@ namespace ledger {
+                result = accumulateBalanceWithOperation(result, *op);
+            }
+
+            return std::make_shared<api::BigIntImpl>(result);
+        }
+
+        api::ERC20LikeBalanceHistory ERC20LikeAccount::getBalanceHistoryFor(
+            const std::chrono::system_clock::time_point& startDate,
+            const std::chrono::system_clock::time_point& endDate,
+            api::TimePeriod period
+        ) {
+            // guard against bad arguments
+            if (startDate > endDate) {
+                throw make_exception(
+                    api::ErrorCode::INVALID_DATE_FORMAT,
+                    "Start date should be strictly greater than end date"
+                );
+            }
+
+            std::vector<std::shared_ptr<api::BigInt>> balances;
+            auto currentBalance = BigInt::ZERO;
+            auto nextDate = DateUtils::incrementDate(startDate, period);
+
+            auto operations = getOperations();
+            // manually sort by date
+            std::sort(
+                operations.begin(),
+                operations.end(),
+                [](
+                    const std::shared_ptr<api::ERC20LikeOperation>& a,
+                    const std::shared_ptr<api::ERC20LikeOperation>& b
+                ) {
+                    return a->getTime() < b->getTime();
                 }
+            );
 
+            // accumulate the balance until we hit the starting date
+            auto opIt = std::cbegin(operations);
+            for (; (*opIt)->getTime() < startDate; ++opIt) {
+                currentBalance = accumulateBalanceWithOperation(currentBalance, **opIt);
+            }
+
+            // iterate over all operations and segment them according to the granularity we’ve
+            // chosen; in our case, we use the default granularity, which is PerDay
+            for (; *opIt; ++opIt) {
+                auto opDate = (*opIt)->getTime();
+
+                // leave the loop if we have hit the upper date bound
+                if (opDate > endDate) {
+                    break;
+                }
+
+                if (opDate >= nextDate) {
+                    balances.push_back(std::make_shared<api::BigIntImpl>(currentBalance));
+                    nextDate = DateUtils::incrementDate(nextDate, period);
+                }
+
+                currentBalance = accumulateBalanceWithOperation(currentBalance, **opIt);
+            }
+
+            // we still have something in the currentBalance that needs to be added to the history
+            // (last day)
+            balances.push_back(std::make_shared<api::BigIntImpl>(currentBalance));
+
+            return api::ERC20LikeBalanceHistory(
+                api::TimePeriod::DAY,
+                startDate,
+                endDate,
+                balances
+            );
+        }
+
+        BigInt ERC20LikeAccount::accumulateBalanceWithOperation(
+            BigInt balance,
```

Comment:
const BigInt& balance, we can avoid additional copy here.

Reply:
The joy of not having a move-as-default language…